File: client/src/execution/logic/executor.rs
//! The `CodeExecutor` FSM (oracle: `shared/execution/CodeExecutor.scala`) — pure
//! `State → State` transitions for one runnable block. The staleness trick: there is no real
//! HTTP cancel, so `cancel`/`started` bump the opaque `RunHandle` and a late result whose
//! handle no longer matches is a NO-OP (`completed`/`failed` return the state unchanged).

use synapse_shared::execution::RunResult;

#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum RunState {
    Idle,
    Running,
    Done,
}

/// Orthogonal to `RunState`; the auth gate is enforced by the CALLER (the identity step), not
/// by the FSM.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub enum EditMode {
    ReadOnly,
    Editing,
}

/// Opaque, monotonic — cannot be fabricated outside this module, so a stored handle can only
/// have come from `started`.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub struct RunHandle(u64);

impl RunHandle {
    const INITIAL: RunHandle = RunHandle(0);

    fn next(self) -> RunHandle {
        RunHandle(self.0 + 1)
    }
}

#[derive(Debug, Clone, PartialEq)]
pub struct ExecutorState {
    pub code: String,
    pub run_state: RunState,
    pub edit_mode: EditMode,
    pub result: Option<RunResult>,
    pub error: Option<String>,
    pub run_id: RunHandle,
}

impl ExecutorState {
    #[must_use]
    pub fn initial(source: &str) -> Self {
        Self {
            code: source.to_owned(),
            run_state: RunState::Idle,
            edit_mode: EditMode::ReadOnly,
            result: None,
            error: None,
            run_id: RunHandle::INITIAL,
        }
    }

    /// Reset-to-starter: identical to `initial` (kept as its own verb, like the oracle).
    #[must_use]
    pub fn reset(source: &str) -> Self {
        Self::initial(source)
    }

    /// A run begins: clear the previous outcome, mint the handle the eventual result must show.
    #[must_use]
    pub fn started(&self) -> Self {
        Self {
            run_state: RunState::Running,
            result: None,
            error: None,
            run_id: self.run_id.next(),
            ..self.clone()
        }
    }

    /// Clear the run outcome (case switch): the buffer and edit unlock survive, the stale
    /// result/error panel disappears, and the bumped handle stale-guards any run in flight —
    /// its reply must not resurrect the panel under the newly selected case.
    #[must_use]
    pub fn clear_outcome(&self) -> Self {
        Self {
            run_state: RunState::Idle,
            result: None,
            error: None,
            run_id: self.run_id.next(),
            ..self.clone()
        }
    }

    /// Cancel without a real HTTP cancel: back to Idle and BUMP the handle, so the in-flight
    /// run's eventual result is stale on arrival.
    #[must_use]
    pub fn cancel(&self) -> Self {
        Self {
            run_state: RunState::Idle,
            run_id: self.run_id.next(),
            ..self.clone()
        }
    }

    /// Apply a result — only if it belongs to the CURRENT run.
    #[must_use]
    pub fn completed(&self, handle: RunHandle, result: RunResult) -> Self {
        if self.run_id != handle {
            return self.clone();
        }
        Self {
            run_state: RunState::Done,
            result: Some(result),
            ..self.clone()
        }
    }

    /// Apply a failure — same staleness guard.
    #[must_use]
    pub fn failed(&self, handle: RunHandle, error: &str) -> Self {
        if self.run_id != handle {
            return self.clone();
        }
        Self {
            run_state: RunState::Done,
            error: Some(error.to_owned()),
            ..self.clone()
        }
    }

    /// Buffer edits touch NOTHING else — a keystroke during a run must not eat the result.
    #[must_use]
    pub fn set_code(&self, code: &str) -> Self {
        Self {
            code: code.to_owned(),
            ..self.clone()
        }
    }

    #[must_use]
    pub fn enter_edit(&self) -> Self {
        Self {
            edit_mode: EditMode::Editing,
            ..self.clone()
        }
    }

    /// Leave edit mode reverting the buffer to the authored source — the last RESULT survives
    /// (reverting code is not un-running it).
    #[must_use]
    pub fn cancel_edit(&self, source: &str) -> Self {
        Self {
            code: source.to_owned(),
            edit_mode: EditMode::ReadOnly,
            ..self.clone()
        }
    }
}

pub fn is_dirty(state: &ExecutorState, source: &str) -> bool {
    state.code != source
}

/// How many lines differ from the authored source, by index (the Edit chrome's badge).
pub fn changed_line_count(state: &ExecutorState, source: &str) -> usize {
    let current: Vec<&str> = state.code.split('\n').collect();
    let authored: Vec<&str> = source.split('\n').collect();
    let max = current.len().max(authored.len());
    (0..max).filter(|&i| current.get(i) != authored.get(i)).count()
}

#[cfg(test)]
#[path = "executor_tests.rs"]
mod tests;

File: web/src/lib/execution/executor.ts
// The `CodeExecutor` FSM (oracle: client/src/execution/logic/executor.rs, itself a port of
// `shared/execution/CodeExecutor.scala`) — pure `State → State` transitions for one runnable
// block. The staleness trick: there is no real HTTP cancel, so `cancel`/`started` bump the
// opaque `RunHandle` and a late result whose handle no longer matches is a NO-OP (`completed`/
// `failed` return the state unchanged).
//
// Functions take the state as their first argument rather than living as methods on a class —
// matching this codebase's other pure-logic ports (`catalog/progress.ts`, `catalog/prefs.ts`),
// where state is plain data and behaviour is free functions over it.

import type { components } from "../api/schema.gen";

type RunResult = components["schemas"]["RunResult"];

export type RunState = "idle" | "running" | "done";

/** Orthogonal to `RunState`; the auth gate is enforced by the CALLER (the identity island), not
 *  by the FSM. */
export type EditMode = "readOnly" | "editing";

// Opaque-ish, monotonic — a branded number rather than a bare one, so a raw number cannot be
// assigned where a handle is expected by accident. The Rust oracle enforces true opacity via
// module privacy (a private tuple field, no public constructor); TS has no equivalent module
// boundary without a WeakMap/closure trick that would be overkill for an ID that only ever needs
// `===` comparison, so this is a deliberate, lighter divergence — same intent, weaker guarantee.
declare const runHandleBrand: unique symbol;
export type RunHandle = number & { readonly [runHandleBrand]: never };

const INITIAL_HANDLE = 0 as RunHandle;

function nextHandle(handle: RunHandle): RunHandle {
  return ((handle as number) + 1) as RunHandle;
}

export interface ExecutorState {
  code: string;
  runState: RunState;
  editMode: EditMode;
  result: RunResult | null;
  error: string | null;
  runId: RunHandle;
}

export function initial(source: string): ExecutorState {
  return {
    code: source,
    runState: "idle",
    editMode: "readOnly",
    result: null,
    error: null,
    runId: INITIAL_HANDLE,
  };
}

/** Reset-to-starter: identical to `initial` (kept as its own verb, like the oracle). */
export function reset(source: string): ExecutorState {
  return initial(source);
}

/** A run begins: clear the previous outcome, mint the handle the eventual result must show. */
export function started(state: ExecutorState): ExecutorState {
  return {
    ...state,
    runState: "running",
    result: null,
    error: null,
    runId: nextHandle(state.runId),
  };
}

/** Clear the run outcome (case switch): the buffer and edit unlock survive, the stale
 *  result/error panel disappears, and the bumped handle stale-guards any run in flight — its
 *  reply must not resurrect the panel under the newly selected case. */
export function clearOutcome(state: ExecutorState): ExecutorState {
  return {
    ...state,
    runState: "idle",
    result: null,
    error: null,
    runId: nextHandle(state.runId),
  };
}

/** Cancel without a real HTTP cancel: back to idle and BUMP the handle, so the in-flight run's
 *  eventual result is stale on arrival. */
export function cancel(state: ExecutorState): ExecutorState {
  return {
    ...state,
    runState: "idle",
    runId: nextHandle(state.runId),
  };
}

/** Apply a result — only if it belongs to the CURRENT run. */
export function completed(state: ExecutorState, handle: RunHandle, result: RunResult): ExecutorState {
  if (state.runId !== handle) return state;
  return {
    ...state,
    runState: "done",
    result,
  };
}

/** Apply a failure — same staleness guard. */
export function failed(state: ExecutorState, handle: RunHandle, error: string): ExecutorState {
  if (state.runId !== handle) return state;
  return {
    ...state,
    runState: "done",
    error,
  };
}

/** Buffer edits touch NOTHING else — a keystroke during a run must not eat the result. */
export function setCode(state: ExecutorState, code: string): ExecutorState {
  return { ...state, code };
}

export function enterEdit(state: ExecutorState): ExecutorState {
  return { ...state, editMode: "editing" };
}

/** Leave edit mode reverting the buffer to the authored source — the last RESULT survives
 *  (reverting code is not un-running it). */
export function cancelEdit(state: ExecutorState, source: string): ExecutorState {
  return { ...state, code: source, editMode: "readOnly" };
}

export function isDirty(state: ExecutorState, source: string): boolean {
  return state.code !== source;
}

/** How many lines differ from the authored source, by index (the Edit chrome's badge). */
export function changedLineCount(state: ExecutorState, source: string): number {
  const current = state.code.split("\n");
  const authored = source.split("\n");
  const max = Math.max(current.length, authored.length);
  let changed = 0;
  for (let i = 0; i < max; i += 1) {
    if (current[i] !== authored[i]) changed += 1;
  }
  return changed;
}
